1. Install and Import Dependencies

In [141]:
!python3 -m pip install paddlepaddle -i https://mirror.baidu.com/pypi/simple

Looking in indexes: https://mirror.baidu.com/pypi/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [165]:
!pip install paddleocr

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [166]:
!git clone https://github.com/PaddlePaddle/PaddleOCR

fatal: destination path 'PaddleOCR' already exists and is not an empty directory.


In [1]:
from paddleocr import PaddleOCR, draw_ocr
from matplotlib import pyplot as plt
import cv2
import os
import requests

2. Applying OCR to images

In [2]:
ocr_model = PaddleOCR(lang='en')

download https://paddleocr.bj.bcebos.com/PP-OCRv3/english/en_PP-OCRv3_det_infer.tar to /Users/hgrover/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer/en_PP-OCRv3_det_infer.tar


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4.00M/4.00M [00:11<00:00, 358kiB/s]


download https://paddleocr.bj.bcebos.com/PP-OCRv3/english/en_PP-OCRv3_rec_infer.tar to /Users/hgrover/.paddleocr/whl/rec/en/en_PP-OCRv3_rec_infer/en_PP-OCRv3_rec_infer.tar


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 9.96M/9.96M [00:21<00:00, 454kiB/s]


download https://paddleocr.bj.bcebos.com/dygraph_v2.0/ch/ch_ppocr_mobile_v2.0_cls_infer.tar to /Users/hgrover/.paddleocr/whl/cls/ch_ppocr_mobile_v2.0_cls_infer/ch_ppocr_mobile_v2.0_cls_infer.tar


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.19M/2.19M [00:09<00:00, 227kiB/s]

[2023/06/26 13:59:24] ppocr DEBUG: Namespace(alpha=1.0, benchmark=False, beta=1.0, cls_batch_num=6, cls_image_shape='3, 48, 192', cls_model_dir='/Users/hgrover/.paddleocr/whl/cls/ch_ppocr_mobile_v2.0_cls_infer', cls_thresh=0.9, cpu_threads=10, crop_res_save_dir='./output', det=True, det_algorithm='DB', det_box_type='quad', det_db_box_thresh=0.6, det_db_score_mode='fast', det_db_thresh=0.3, det_db_unclip_ratio=1.5, det_east_cover_thresh=0.1, det_east_nms_thresh=0.2, det_east_score_thresh=0.8, det_limit_side_len=960, det_limit_type='max', det_model_dir='/Users/hgrover/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer', det_pse_box_thresh=0.85, det_pse_min_area=16, det_pse_scale=1, det_pse_thresh=0, det_sast_nms_thresh=0.2, det_sast_score_thresh=0.5, draw_img_save_dir='./inference_results', drop_score=0.5, e2e_algorithm='PGNet', e2e_char_dict_path='./ppocr/utils/ic15_dict.txt', e2e_limit_side_len=768, e2e_limit_type='max', e2e_model_dir=None, e2e_pgnet_mode='fast', e2e_pgnet_score_thresh=0.5, e

In [6]:
img_path = os.path.join('.', '../static/drug1.jpg')

In [7]:
img_path

'./../static/drug1.jpg'

In [8]:
result = ocr_model.ocr(img_path)

[2023/06/26 14:00:14] ppocr WARNING: Since the angle classifier is not initialized, the angle classifier will not be uesd during the forward process
[2023/06/26 14:00:14] ppocr DEBUG: dt_boxes num : 57, elapse : 0.4027590751647949
[2023/06/26 14:00:19] ppocr DEBUG: rec_res num  : 57, elapse : 5.045177698135376


In [9]:
result

[[[[[22.0, 18.0], [191.0, 23.0], [190.0, 46.0], [21.0, 40.0]],
   ('VCVS pharmacy', 0.9200306534767151)],
  [[[276.0, 15.0], [368.0, 15.0], [368.0, 39.0], [276.0, 39.0]],
   ('PATIENT', 0.9978836178779602)],
  [[[423.0, 12.0], [485.0, 10.0], [486.0, 36.0], [424.0, 39.0]],
   ('John', 0.9999155402183533)],
  [[[760.0, 17.0], [855.0, 17.0], [855.0, 35.0], [760.0, 35.0]],
   ('Important', 0.9887751340866089)],
  [[[428.0, 47.0], [477.0, 47.0], [477.0, 69.0], [428.0, 69.0]],
   ('Doe', 0.9981111884117126)],
  [[[758.0, 38.0], [871.0, 38.0], [871.0, 57.0], [758.0, 57.0]],
   ('Information', 0.9947527050971985)],
  [[[22.0, 48.0], [109.0, 48.0], [109.0, 62.0], [22.0, 62.0]],
   ('123 Main Street', 0.9680683612823486)],
  [[[177.0, 48.0], [213.0, 48.0], [213.0, 62.0], [177.0, 62.0]],
   ('#0000', 0.9995212554931641)],
  [[[158.0, 62.0], [213.0, 62.0], [213.0, 76.0], [158.0, 76.0]],
   ('cvs.com', 0.9069531559944153)],
  [[[246.0, 53.0], [394.0, 51.0], [394.0, 72.0], [246.0, 74.0]],
   ('METFO

In [10]:
for resu in result:
  for res in resu:
    print(res[-1])

('VCVS pharmacy', 0.9200306534767151)
('PATIENT', 0.9978836178779602)
('John', 0.9999155402183533)
('Important', 0.9887751340866089)
('Doe', 0.9981111884117126)
('Information', 0.9947527050971985)
('123 Main Street', 0.9680683612823486)
('#0000', 0.9995212554931641)
('cvs.com', 0.9069531559944153)
('METFORMIN HCL', 0.9821059107780457)
('Anytown,USA 00000', 0.9456605315208435)
('500 MG', 0.9479793906211853)
('Talk to your Dr.about', 0.9553543925285339)
('Store Corporate Name', 0.9770164489746094)
('300 East Street,Anytown,USA 00000', 0.9481533169746399)
('the safe use of alcohol', 0.9440683126449585)
('TEL555-555-5555', 0.9943721890449524)
('METFORMIN HCL', 0.9616110324859619)
('while taking this drug', 0.9653608798980713)
('1', 0.998582124710083)
('RX 105004', 0.9216597676277161)
('Severe vomiting or', 0.9760987758636475)
('TABLET', 0.9972255825996399)
('500 MG TABLET', 0.938424825668335)
('MORNING', 0.9972584843635559)
('diarrhea may cause', 0.97607421875)
('QTY:60', 0.935261011123657

The below is an OCR scan of a prescription label. Create a python tuple with the following information; Medication name, how many pills to take in a day (Written as "_ pills per day"), how many pills to take for each occurrence, Prescription date (numbers only), Expiration date (numbers only). Provide only the tuple, no other text.

('VCVS pharmacy', 0.9200306534767151)
('PATIENT', 0.9978834986686707)
('John', 0.9999155402183533)
('Important', 0.9887750148773193)
('Doe', 0.9981111884117126)
('Information', 0.9947527050971985)
('123 Main Street', 0.9680683612823486)
('#0000', 0.9995212554931641)
('cvs.com', 0.9069531559944153)
('METFORMIN HCL', 0.9821059703826904)
('Anytown,USA 00000', 0.9456606507301331)
('500 MG', 0.9479793906211853)
('Talk to your Dr.about', 0.9553545117378235)
('Store Corporate Name', 0.9770164489746094)
('300 East Street,Anytown,USA 00000', 0.9481532573699951)
('the safe use of alcohol', 0.9440683126449585)
('TEL555-555-5555', 0.9943720698356628)
('METFORMIN HCL', 0.9616110324859619)
('while taking this drug', 0.9653609991073608)
('1', 0.998582124710083)
('RX 105004', 0.9216598868370056)
('Severe vomiting or', 0.9760987758636475)
('TABLET', 0.9972255825996399)
('500 MG TABLET', 0.9384247064590454)
('MORNING', 0.9972584843635559)
('diarrhea may cause', 0.9760740995407104)
('QTY:60', 0.9352610111236572)
('dehvdration if these', 0.9557657241821289)
('occur call MD', 0.9598146677017212)
('REFILLS: 2 by MM/DD/YYYY', 0.967379093170166)
('Drug may cause', 0.9538489580154419)
('PRSCBR:J.Doe', 0.9414737820625305)
('Generic equivalent of:Glucophage', 0.9658290147781372)
('MIDDAY', 0.9973875880241394)
('Take 1 tablet orally', 0.9962593913078308)
('lactic acidosis.If', 0.9091899394989014)
('DATE FILLED:06/26/2023', 0.9532528519630432)
('symptoms develop.', 0.9701091647148132)
('DISCARDAFTER06/26/2025', 0.989685595035553)
('2 times a day', 0.9771130084991455)
('seek medical help.', 0.9660184383392334)
('RPH:J.Smith', 0.9579192996025085)
('MFR:ABC Manufacturer', 0.9491662979125977)
('EVENING', 0.9966979622840881)
('TABLET', 0.9974088072776794)
('This is a WHITE,OVAL-SHAPED', 0.939594030380249)
('tablet imprinted with 93 on the front', 0.9532434344291687)
('CAUTION Federal law prohibits the', 0.9679855704307556)
('and 48 on the back.', 0.9214495420455933)
('REFILL', 0.998106062412262)
('BEDTIME', 0.9972743391990662)
('transfer of this drug to any person', 0.9546133875846863)
('other than the patient for whom it', 0.9789407253265381)
('APHARMACY ADVICE', 0.9722907543182373)
('was prescribed.', 0.927888035774231)

The below is an OCR scan of a prescription label. Create a python tuple with the following information; Medication name, how many pills to take in a day (Written as "_ pills per day"), how many pills to take for each occurrence, Prescription date (numbers only), Expiration date (numbers only). Provide only the tuple, no other text.


The below is an OCR scan of a prescription label. Create a python tuple with the following information; Medication name, how many pills to take in a day (Written as "_ pills per day"), how many pills to take for each occurrence, Prescription date (numbers only), Expiration date (numbers only). Provide only the tuple, no other text. 

('VCVS pharmacy', 0.9200306534767151)
('PATIENT', 0.9978836178779602)
('John', 0.9999155402183533)
('Important', 0.9887751340866089)
('Doe', 0.9981111884117126)
('Information', 0.9947527050971985)
('123 Main Street', 0.9680683612823486)
('#0000', 0.9995212554931641)
('cvs.com', 0.9069531559944153)
('METFORMIN HCL', 0.9821059107780457)
('Anytown,USA 00000', 0.9456605315208435)
('500 MG', 0.9479793906211853)
('Talk to your Dr.about', 0.9553543925285339)
('Store Corporate Name', 0.9770164489746094)
('300 East Street,Anytown,USA 00000', 0.9481533169746399)
('the safe use of alcohol', 0.9440683126449585)
('TEL555-555-5555', 0.9943721890449524)
('METFORMIN HCL',

In [21]:


def get_completion(prompt, model="gpt-3.5-turbo"):
    messages = [{"role": "user", "content": prompt}]
    response = openai.ChatCompletion.create(
        model=model,
        messages=messages,
        temperature=0, # this is the degree of randomness of the model's output
    )
    return response.choices[0].message["content"]


ocr_extract = []
for resu in result:
    for res in resu:
        ocr_extract.append(str(res[-1]))

ocr_extract_str = '\n'.join(ocr_extract) 
prompt = f'The below is an OCR scan of a prescription label. Create a python tuple with the following information; Medication name, how many pills to take in a day (Written as "_ pills per day"), how many pills to take for each occurrence, Prescription date (numbers only), Expiration date (numbers only). Provide only the tuple, no other text. \n\n{ocr_extract_str}'
print(prompt)


The below is an OCR scan of a prescription label. Create a python tuple with the following information; Medication name, how many pills to take in a day (Written as "_ pills per day"), how many pills to take for each occurrence, Prescription date (numbers only), Expiration date (numbers only). Provide only the tuple, no other text. 

('VCVS pharmacy', 0.9200306534767151)
('PATIENT', 0.9978836178779602)
('John', 0.9999155402183533)
('Important', 0.9887751340866089)
('Doe', 0.9981111884117126)
('Information', 0.9947527050971985)
('123 Main Street', 0.9680683612823486)
('#0000', 0.9995212554931641)
('cvs.com', 0.9069531559944153)
('METFORMIN HCL', 0.9821059107780457)
('Anytown,USA 00000', 0.9456605315208435)
('500 MG', 0.9479793906211853)
('Talk to your Dr.about', 0.9553543925285339)
('Store Corporate Name', 0.9770164489746094)
('300 East Street,Anytown,USA 00000', 0.9481533169746399)
('the safe use of alcohol', 0.9440683126449585)
('TEL555-555-5555', 0.9943721890449524)
('METFORMIN HCL',

In [44]:
import os
import openai
# openai.api_key = "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"
openai.api_key = "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"


rv = get_completion(prompt)

In [45]:
rv

"('METFORMIN HCL', '2 pills per day', '1 pill', '06262023', '06262025')"

In [28]:
type(eval(rv))

tuple

In [46]:
import datetime
import re

rv = list(eval(rv))
rv[-2] = datetime.datetime.strptime(rv[-2], '%m%d%Y')
rv[-1] = datetime.datetime.strptime(rv[-1], '%m%d%Y')

rv[1] = int(re.findall('[0-9]+', rv[1])[0])
rv[2] = int(re.findall('[0-9]+', rv[2])[0])
print(rv)

['METFORMIN HCL', 2, 1, datetime.datetime(2023, 6, 26, 0, 0), datetime.datetime(2025, 6, 26, 0, 0)]


In [31]:
rv

['(',
 "'",
 'M',
 'E',
 'T',
 'F',
 'O',
 'R',
 'M',
 'I',
 'N',
 ' ',
 'H',
 'C',
 'L',
 "'",
 ',',
 ' ',
 "'",
 '2',
 ' ',
 'p',
 'i',
 'l',
 'l',
 's',
 ' ',
 'p',
 'e',
 'r',
 ' ',
 'd',
 'a',
 'y',
 "'",
 ',',
 ' ',
 "'",
 '1',
 ' ',
 'p',
 'i',
 'l',
 'l',
 "'",
 ',',
 ' ',
 "'",
 '0',
 '6',
 '2',
 '6',
 '2',
 '0',
 '2',
 '3',
 "'",
 ',',
 ' ',
 "'",
 '0',
 '6',
 '2',
 '6',
 '2',
 '0',
 '2',
 '5',
 "'",
 ')']